extensions/llama/typescript/llama.ts

```diff
@@ -0,0 +1,343 @@
+const inclusion = require("inclusion");
+
+const DEFAULT_MODEL_DIR = path.join(process.cwd(), "models");
+
+export const SUPPORTED_MODELS: LlamaModelName[] = ["llama-2-7b-chat"];
+
+export type ModelOptions = Omit<LlamaModelOptions, "modelPath">;
+export type LlamaModelName = "llama-2-7b-chat";
+
+export interface LlamaCompletionParams extends LLamaChatPromptOptions {
+  model: LlamaModelName;
+  prompt: string;
+
+  // Used when constructing the session to perform the prompt
+  conversationHistory?: ConversationInteraction[];
+  systemPrompt?: string;
+}
+
+/**
+ * A model parser for LLaMA models, leveraging llama-node API. Construct this parser with the absolute
+ * path to the directory containing the LLaMA model GGUF format (.gguf) files for the supported models.
+ * The file names must contain the LlamaModelName, e.g. "llama-2-7b-chat.ggmlv3.q4_K_M.bin".
+ * e.g. curl -L https://huggingface.co/TheBloke/Llama-2-7B-Chat-GGUF/resolve/main/llama-2-7b-chat.Q4_K_M.gguf --output ./models/llama-2-7b-chat.Q4_K_M.gguf
+ */
+export class LlamaModelParser extends ParameterizedModelParser<LlamaCompletionParams> {
+  _id = "LLaMA";
+
+  modelDir: string = DEFAULT_MODEL_DIR;
+  modelOptions?: ModelOptions;
+
+  private modelContexts: Map<LlamaModelName, LlamaContext> = new Map();
+
+  public constructor(modelDir?: string, modelOptions?: ModelOptions) {
+    super();
+    this.modelDir = modelDir ?? this.modelDir;
+    this.modelOptions = modelOptions;
+  }
+
+  private async getModelSession(
+    model: LlamaModelName,
+    systemPrompt?: string,
+    conversationHistory?: ConversationInteraction[]
+  ): Promise<LlamaChatSession> {
+    if (!this.modelContexts.has(model)) {
+      // Need to dynamic import ES module to use in our CommonJS module
+      const { LlamaModel, LlamaContext } = await inclusion("node-llama-cpp");
+
+      let modelPath;
+
+      const modelFiles = await readdir(this.modelDir);
+      for (const file of modelFiles) {
+        if (path.extname(file) === ".gguf" && file.includes(model)) {
+          modelPath = path.join(this.modelDir, file);
+          break;
+        }
+      }
+
+      if (!modelPath) {
+        throw new Error(`Could not find model ${model} in ${this.modelDir}`);
+      }
+
+      const llamaModel: LlamaModel = new LlamaModel({
+        ...this.modelOptions,
+        modelPath,
+      });
+
+      const context: LlamaContext = new LlamaContext({ model: llamaModel });
+      this.modelContexts.set(model, context);
+    }
+
+    const context = this.modelContexts.get(model)!;
+    const { LlamaChatSession } = await inclusion("node-llama-cpp");
+
+    return new LlamaChatSession({ context, conversationHistory, systemPrompt });
+  }
+
+  public serialize(
+    promptName: string,
+    data: LlamaCompletionParams,
+    aiConfig: AIConfigRuntime,
+    params?: JSONObject | undefined
+  ): Prompt | Prompt[] {
+    const { prompt: input, conversationHistory, ...completionParams } = data;
```

Comment:
Looks like `conversationHistory` isn't serialized -- should that get saved each as separate Prompt objects? I believe that's what the openai model parser does.

Reply:
Hmm, I'm not sure -- this is where the serialize/deserialize is confusing. The Llama model would never have `conversationHistory` as a param to its completion request and is only something constructed by the parser at run time based on existing prompts. I can try to add that using openai as an example, but seems a bit strange imo

Reply:
Done